File: 0141_linked_list_cycle/hasCycle.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def hasCycle(head: ListNode) -> bool:
    """
    Time: O(n)
    Space: O(1)
    """
    tort = hare = head
    while hare and hare.next:
        tort = tort.next
        hare = hare.next.next
        if tort == hare:
            return True
    return False

File: 0141_linked_list_cycle/hasCycle.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

public class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null) {
            return false;
        }
        ListNode tort = head, hare = head;
        while (hare.next != null && hare.next.next != null) {
            hare = hare.next.next;
            tort = tort.next;
            if (hare == tort) {
                return true;
            }
        }
        return false;
    }
}